[SPARK-11373] [CORE] Add metrics to the History Server and FsHistoryProvider

steveloughran opened the pull request:
This adds metrics to the history server, with the `FsHistoryProvider` metering its load, performance and reliability.

see [SPARK-11373](https://issues.apache.org/jira/browse/SPARK-11373)

The `HistoryServer` sets up the codahale metrics for the Web under metrics/ with metrics/metrics behind metrics, metrics/health any health probes and metrics/threads a thread dump. There's currently no attempt to  hook up JMX, etc. The Web servlets are the ones tests can easily hit and don't need infrastructure, so are the good initial first step.

It then passes the metrics and health registries down to the providers in a `ApplicationHistoryBinding` case class, via a new method

```
def start(binding: ApplicationHistoryBinding): Unit
```

The base class has implementation so that all existing providers will still link properly; the base implementation currently checks and fails
 the use of a binding case class is also to ensure that if new binding information were added in future, existing implementations would still link.

The `FsHistoryProvider` implements the `start()` method, registering two counters and two timers.
1. Number of update attempts and number of failed updates —and the same for app UI loads.
2. Time for updates and app UI loads.

Points of note
- Why not use Spark's `MetricsSystem`? I did start off with that, but it needs a `SparkContext` to run off, which the server doesn't have. Ideally that would be way to go, as it would support all the spark conf -based metrics setup. Someone who understands the `MetricsSystem` would need to get involved here as would make for a more complex patch. In `FsHistoryProvider` the registry information is all kept in a `Source` subclass for ease of future migration to `MetricsSystem`.
- Why the extra `HealthRegistry`? It's a nice way of allowing providers to indicate (possibly transient) health problems for monitoring tools/clients to hit. For the FS provider it could maybe flag when there hadn't been any successful update for a specified time period. (that could also be indicated by having a counter of "seconds since last update" and let monitoring tools monitor the counter value and act on it). Access control problems to the directory is something else which may be considered a liveness problem: it won't get better without human intervention
- The `FsHistoryProvider.start()` method should really take the thread start code from from class constructor's `initialize()` method. This would ensure that incomplete classes don't get called by spawned threads, and makes it possible for test-time subclasses to skip thread startup. I've not attempted to do that in this patch.
- No tests for this yet. Hitting the three metrics servlets in the HistoryServer is the obvious route; the JSON payload of the metrics can be parsed and scanned for relevant counters too. 
- Part of the patch for `HistoryServerSuite` removes the call to `HistoryServer.initialize()` the `before` clause. That was a duplicate call, one which hit the re-entrancy tests on the provider & registry. As well as cutting it, `HistoryServer.initialize()` has been made idempotent. That should not be needed -but it will eliminate the problem arising again.

Once the SPARK-1537 YARN timeline server history provider is committed, then I'll add metrics support there too. The YARN timeline provider would:
1. Add timers of REST operations as well as playback load times, which can count network delays as well as JSON deserialization overhead. 
2. Add a health check for connectivity too: the timeline server would be unhealthy if connections to the timeline server were either blocking or failing. And again, if there were security/auth problems, they'd be considered non-recoverable.
3. Move thread launch under the `start()` method, with some test subclasses disabling thread launch.

steveloughran committed:
[SPARK-11373] First pass at adding metrics to the history server, with the FsHistoryProvider counting
1. Number of update attempts and number of failed updates
2. Time for updates and app UI loads

The HistoryServer sets up the codahale metrics for the Web under metrics/ with metrics/metrics behind metrics, metrics/health any health probes and metrics/threads a thread dump.

steveloughran committed:
[SPARK-11373] tests and review; found and fixed a re-entrancy in HistoryServerSuite which was causing problems with counter registration

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45386 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45386/consoleFull)** for PR 9571 at commit [`cb7cddb`](https://github.com/apache/spark/commit/cb7cddb0443d4d8fac7d63e604fe303f5f57d5cd).

SparkQA commented:
**[Test build #45386 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45386/consoleFull)** for PR 9571 at commit [`cb7cddb`](https://github.com/apache/spark/commit/cb7cddb0443d4d8fac7d63e604fe303f5f57d5cd).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

steveloughran committed:
[SPARK-11373] scala-style javadoc tuning in FsHistoryProvider

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

CharlesYeh commented:
Thanks! I'll look into this

CharlesYeh commented:
I'm assuming you didn't make a HistorySource in HistoryServer (what MasterSource in Master does) because it requires the MetricsSystem.

Where does MetricsSystem require a SparkContext? It looks like it just takes SparkConf which HistoryServer has.

steveloughran commented:
...I thought it did. If I'm wrong, then yes, all you need is a context. Once that's in there, the HistorySource should be what's needed

steveloughran commented on core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala:
```diff
@@ -121,20 +132,36 @@ class HistoryServer(
+
+      // hook up Codahale metrics
+      val metricsHandler = new ServletContextHandler(null, HistoryServer.METRICS_PATH_PREFIX)
+      metricsHandler.addServlet(new ServletHolder(new HealthCheckServlet(health)),
```
given the repetition here, this could take a list of (servlet, path) and apply that to the registration

steveloughran commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -678,6 +709,69 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+      def register(s: String, m: Metric) = {
+        metricRegistry.register(s, m)
+      }
+      register("update.count", updateCount)
```
copy design in #6935 and use a list of (name, metric) both for registration & generating toString value

steveloughran committed:
SPARK-11373: move to MetricsSystem, though retaining/expanding health checks, which aren't part of it

SparkQA commented:
**[Test build #46779 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46779/consoleFull)** for PR 9571 at commit [`f6bf558`](https://github.com/apache/spark/commit/f6bf5587f6c76753a1e161bbae0050a49bf6d872).

steveloughran committed:
scalacheck

SparkQA commented:
**[Test build #46779 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46779/consoleFull)** for PR 9571 at commit [`f6bf558`](https://github.com/apache/spark/commit/f6bf5587f6c76753a1e161bbae0050a49bf6d872).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46779/
Test FAILed.

steveloughran commented:
Added reworked design

* all metrics go through the `MetricsSystem`; the providers return an optional `Source` from the `start()` call.
* `FsHistoryProvider` metrics split lookups of missing files from failed attempts to replay the logs.
* `FsHistoryProvider` metrics include time of the mergeListing() operation, which can be quite the CPU killer.
* There's a `HealthSource` for health checks; it's being explicitly managed in the HistoryServer.
* And there's an initial health check for the FS, which simply returns FS safe/mode flag. 

Really, the health check logic needs its own `HealthSystem` for the register/unregister. Trying to design one that spans all the applications is more complex and I'm trying to avoid that.

Furthermore, those operations which fail asynchronous and just have exceptions logged should have their exceptions saved to another health check (I may implement that). That way the health check will react to live system failures, including things like the log directory being deleted during a run.

steveloughran commented:
jenkins, test this please

SparkQA commented:
**[Test build #46780 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46780/consoleFull)** for PR 9571 at commit [`1dcbb5f`](https://github.com/apache/spark/commit/1dcbb5fce89a5f5f9a19b846432ac4d9937a23d0).

SparkQA commented:
**[Test build #46781 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46781/consoleFull)** for PR 9571 at commit [`1dcbb5f`](https://github.com/apache/spark/commit/1dcbb5fce89a5f5f9a19b846432ac4d9937a23d0).

steveloughran committed:
SPARK-11373: cut out the notion of binding information; simplify provider launch

SparkQA commented:
**[Test build #46782 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46782/consoleFull)** for PR 9571 at commit [`25e77bd`](https://github.com/apache/spark/commit/25e77bd8eb1d8908410802dc2c59274b15f7294b).

steveloughran commented:
(Note that the POMs changed to pull in some more of the codahale servlets, though only the health checks & thread dump are being registered. Hooking up those servlets through the MetricsSystem may be a bit tricky; and as it has its own metrics, is probably the one to go for. In which case: wrapping up JVM stats as a `Source` would give automatic access to those numbers

SparkQA commented:
**[Test build #46782 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46782/consoleFull)** for PR 9571 at commit [`25e77bd`](https://github.com/apache/spark/commit/25e77bd8eb1d8908410802dc2c59274b15f7294b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46782/
Test FAILed.

SparkQA commented:
**[Test build #46780 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46780/consoleFull)** for PR 9571 at commit [`1dcbb5f`](https://github.com/apache/spark/commit/1dcbb5fce89a5f5f9a19b846432ac4d9937a23d0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46780/
Test PASSed.

SparkQA commented:
**[Test build #46781 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46781/consoleFull)** for PR 9571 at commit [`1dcbb5f`](https://github.com/apache/spark/commit/1dcbb5fce89a5f5f9a19b846432ac4d9937a23d0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46781/
Test PASSed.

steveloughran committed:
SPARK-11373 cut the health checks out

SparkQA commented:
**[Test build #47440 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47440/consoleFull)** for PR 9571 at commit [`2e04803`](https://github.com/apache/spark/commit/2e04803528c9820ac1e47a8470d7e34cdd98faff).

steveloughran commented:
latest branch cut out health checks. They'd be nice, but as they are useful across all of spark, it's probably best to wait for something central to go in and pick it up, rather than try and squeeze something in here.

SparkQA commented:
**[Test build #48567 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48567/consoleFull)** for PR 9571 at commit [`06e5289`](https://github.com/apache/spark/commit/06e5289fae86b3d12102bf018cc301da9964e0d0).

SparkQA commented:
**[Test build #48567 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48567/consoleFull)** for PR 9571 at commit [`06e5289`](https://github.com/apache/spark/commit/06e5289fae86b3d12102bf018cc301da9964e0d0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48567/
Test PASSed.

SparkQA commented:
**[Test build #48768 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48768/consoleFull)** for PR 9571 at commit [`d6fa568`](https://github.com/apache/spark/commit/d6fa568fab72a2c4d57ecfcd304d000379534990).

SparkQA commented:
**[Test build #48768 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48768/consoleFull)** for PR 9571 at commit [`d6fa568`](https://github.com/apache/spark/commit/d6fa568fab72a2c4d57ecfcd304d000379534990).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48768/
Test PASSed.

SparkQA commented:
**[Test build #50202 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50202/consoleFull)** for PR 9571 at commit [`984100d`](https://github.com/apache/spark/commit/984100d39f9ba137a9ca4dc2a06ce6393f21ad6a).

SparkQA commented:
**[Test build #50202 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50202/consoleFull)** for PR 9571 at commit [`984100d`](https://github.com/apache/spark/commit/984100d39f9ba137a9ca4dc2a06ce6393f21ad6a).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50202/
Test FAILed.

steveloughran committed:
[SPARK-11373] tail end of rebase operation

SparkQA commented:
**[Test build #50822 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50822/consoleFull)** for PR 9571 at commit [`3ceeb9e`](https://github.com/apache/spark/commit/3ceeb9ef2cd34d5d5eeccacd9e49477c133f0ee0).

SparkQA commented:
**[Test build #50822 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50822/consoleFull)** for PR 9571 at commit [`3ceeb9e`](https://github.com/apache/spark/commit/3ceeb9ef2cd34d5d5eeccacd9e49477c133f0ee0).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50822/
Test FAILed.

steveloughran committed:
[SPARK-11373] scalastyle and import ordering

SparkQA commented:
**[Test build #50921 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50921/consoleFull)** for PR 9571 at commit [`889978c`](https://github.com/apache/spark/commit/889978c52a9efa29e287d6e6d4c57f1fa9717442).

SparkQA commented:
**[Test build #50921 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50921/consoleFull)** for PR 9571 at commit [`889978c`](https://github.com/apache/spark/commit/889978c52a9efa29e287d6e6d4c57f1fa9717442).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50921/
Test FAILed.

steveloughran committed:
SPARK-11373 finish review of merge with trunk; add new Timestamp gauge and set to time of update/update success. Makes update time visible for tests/users/ops

SparkQA commented:
**[Test build #51188 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51188/consoleFull)** for PR 9571 at commit [`de1769b`](https://github.com/apache/spark/commit/de1769b85b7a9ac3fc160eaf7d465176e95d7100).

SparkQA commented:
**[Test build #51188 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51188/consoleFull)** for PR 9571 at commit [`de1769b`](https://github.com/apache/spark/commit/de1769b85b7a9ac3fc160eaf7d465176e95d7100).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51188/
Test FAILed.

steveloughran committed:
[SPARK-11373] finish rebasing to master, correct tightened style checks

SparkQA commented:
**[Test build #57010 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57010/consoleFull)** for PR 9571 at commit [`a5d1f4c`](https://github.com/apache/spark/commit/a5d1f4c3d1d7076ad7c65343d9c33beaab71afbe).

SparkQA commented:
**[Test build #57010 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57010/consoleFull)** for PR 9571 at commit [`a5d1f4c`](https://github.com/apache/spark/commit/a5d1f4c3d1d7076ad7c65343d9c33beaab71afbe).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57010/
Test PASSed.

SparkQA commented:
**[Test build #58471 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58471/consoleFull)** for PR 9571 at commit [`a5d1f4c`](https://github.com/apache/spark/commit/a5d1f4c3d1d7076ad7c65343d9c33beaab71afbe).

SparkQA commented:
**[Test build #58471 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58471/consoleFull)** for PR 9571 at commit [`a5d1f4c`](https://github.com/apache/spark/commit/a5d1f4c3d1d7076ad7c65343d9c33beaab71afbe).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58471/
Test FAILed.

SparkQA commented:
**[Test build #58563 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58563/consoleFull)** for PR 9571 at commit [`64de389`](https://github.com/apache/spark/commit/64de38964c05b514e921c15962be41e92519e98f).

SparkQA commented:
**[Test build #58566 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58566/consoleFull)** for PR 9571 at commit [`1b276e4`](https://github.com/apache/spark/commit/1b276e441dabab2ac20a98c03de01975c95feb72).

SparkQA commented:
**[Test build #58563 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58563/consoleFull)** for PR 9571 at commit [`64de389`](https://github.com/apache/spark/commit/64de38964c05b514e921c15962be41e92519e98f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58563/
Test PASSed.

SparkQA commented:
**[Test build #58566 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58566/consoleFull)** for PR 9571 at commit [`1b276e4`](https://github.com/apache/spark/commit/1b276e441dabab2ac20a98c03de01975c95feb72).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58566/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala:
```diff
@@ -114,28 +123,45 @@ class HistoryServer(
    */
   def initialize() {
-    attachPage(new HistoryPage(this))
+    if (!initialized.getAndSet(true)) {
```
In which cases is this method called twice?

steveloughran replied:
I don't remember now, but it did happen.

steveloughran replied:
It's in `HistoryServerSuite`

```scala
server = new HistoryServer(conf, provider, securityManager, 18080)
server.initialize()
server.bind()
```

issue here is the HS constructor already calls `initialize()`; some of the tests call it twice.

One option here would be move that `initialize()` out of the constructor of HS; that'd mean changing `main()`, but produce something better for subclassing and testing anyway. Even so, making `initialize()`  non-reentrant can only be a good thing, given that threads are spawned off, metrics registered, etc.

vanzin replied:
If that's the case, `initialize()` should be made private and its calls removed from the tests.

(I've seen this pattern of calling an `init()` method from constructors in Scala when the initialization code requires local variables, which otherwise would have to be fields in the class...)

steveloughran replied:
...Pulled it from the test suite, so the check is superflous, and cut it...

Looking at this more, I think the metrics setup should be moved to `bind()`, rather than `initialize()`, precisely because the metrics setup may be doing things in terms of thread startup, and you don't want that to happen so early in class construction. Doing that and the testing/review to make sure that the later-metrics-init won't break assumptions elsewhere.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -278,6 +303,9 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+    metrics.updateCount.inc()
+    metrics.updateLastAttempted.touch()
+    time(metrics.updateTimer) {
     try {
```
Don't you need to indent this block of code now?

steveloughran replied:
done

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -667,6 +700,90 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+  val appUILoadNotFoundCount = new Counter()
+
+  /** Statistics on time to load app UIs. */
+  val appUiLoadTimer = new Timer()
```
Given that different apps can have very different event logs, I wonder how useful this metric really is? Perhaps some extra processing needs to be done (e.g. events processed per sec or something)?

steveloughran replied:
good point. That'd be a nice throughput. all that's needed is to count the #of events loaded and divide them for a playback rate, some events/second or nanos/event. I'd prefer the latter for any microbenchmarking; the events/second metric probably better for some monitoring display.

vanzin replied:
I see you added the more useful `appUIEventReplayTime`, so is this metric needed? I don't think a lot of useful information, if any, can be derived from it.

vanzin replied:
Ping. I see you use this in `time()` to update the total time, but do you need to register this metric below?

steveloughran replied:
I've cut the metric

steveloughran committed:
[SPARK-11373] Address review comments and add a new counter of events played back during merge and U load

SparkQA commented:
**[Test build #60296 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60296/consoleFull)** for PR 9571 at commit [`5eb5360`](https://github.com/apache/spark/commit/5eb536023c0d363728911487942d47fec2e99fb7).

steveloughran commented:
Updated patch. Addresses indentation, found and eliminated one more call to {{initialize()}} outside of constructor.

Adds a whole new counter, `event.replay.count`, which counts the number of events replayed during load or attempt merge. I'm not trying to be clever about deriving playback rate or anything, as it could get complex with the merge/update listings being included. However, including those playback events could be useful in identifying why startup was slow, or why updates are taking a while to surface.

SparkQA commented:
**[Test build #60296 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60296/consoleFull)** for PR 9571 at commit [`5eb5360`](https://github.com/apache/spark/commit/5eb536023c0d363728911487942d47fec2e99fb7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60296/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/ApplicationHistoryProvider.scala:
```diff
@@ -17,9 +17,12 @@
 
 import org.apache.spark.SparkException
+import org.apache.spark.metrics.source.Source
+import org.apache.spark.scheduler.{SparkListener, SparkListenerApplicationEnd, SparkListenerApplicationStart, SparkListenerBlockManagerAdded, SparkListenerBlockManagerRemoved, SparkListenerBlockUpdated, SparkListenerEnvironmentUpdate, SparkListenerEvent, SparkListenerExecutorAdded, SparkListenerExecutorMetricsUpdate, SparkListenerExecutorRemoved, SparkListenerJobEnd, SparkListenerJobStart, SparkL ...
```
At this point I'd just import `org.apache.spark.scheduler._`.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/ApplicationHistoryProvider.scala:
```diff
@@ -110,3 +127,87 @@ private[history] abstract class ApplicationHistoryProvider {
+ * A simple counter of events.
+ * There is no concurrency support here: all events must come in sequentially.
+ */
+private[history] class EventCountListener extends SparkListener {
```
Isn't it easier to extend `SparkFirehoseListener` if all you're doing is counting?

steveloughran replied:
I was sure there was something to do this, I just couldn't find it when I looked. will do

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -278,55 +304,63 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
-              logError("Exception while merging application listings", e)
+        logInfos.map { file =>
+            replayExecutor.submit(new Runnable {
+              override def run(): Unit = time(metrics.mergeApplicationListingTimer) {
```
This metric contains the time to replay an app event log, which can vary a lot per application; same issue with the UI load time issue I pointed out before. That makes this metric not very useful.

vanzin replied:
I commented later on the event count metric update; but this would be more useful if you knew the event count related to it, and if it were more targeted (i.e. just the replay, not including the time taken to merge the internal lists).

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -395,7 +429,8 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
         val bus = new ReplayListenerBus()
-        val res = replay(fileStatus, bus)
+        val (res, count) = replay(fileStatus, bus)
+        metrics.eventReplayedCount.inc(count)
```
You're using this metric in two different contexts: while creating the listing, and while building the UI.

That means there's no good way to get a useful metric out of it, since you don't know how many events correspond to each and you're measuring the times for each separately.

steveloughran replied:
I will measure them both then, and add some time-to-load metric

steveloughran commented:
One other metric set I'm thinking of relates to a JIRA on app UIs not being visible: making the time of last scan a metric, both as an epoch time and diff from current time. That would let a management tool keep track of the scan history, maybe even show alerts if things are out of date. At the very least, if there are support calls, *and someone were to add the coda hale metrics servlets* then users could point their browsers at the metrics page and get a lump of JSON to paste into JIRAs

steveloughran committed:
[SPARK-11373] more metrics, all sources have prefixes, common features pulled up into HistoryMetricSource, including registration and prefix support. The existing test case "incomplete apps get refreshed" has been extended to check for metrics in the listings, alongside some specific probes for values in the fs history provider (loads, load time average > 0)

SparkQA commented:
**[Test build #62318 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62318/consoleFull)** for PR 9571 at commit [`6e7f466`](https://github.com/apache/spark/commit/6e7f466da0681af2c40a6842ed6c542ca44c7bd5).

SparkQA commented:
**[Test build #62318 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62318/consoleFull)** for PR 9571 at commit [`6e7f466`](https://github.com/apache/spark/commit/6e7f466da0681af2c40a6842ed6c542ca44c7bd5).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62318/
Test FAILed.

steveloughran committed:
[SPARK-11373] add a check that before there's been any events loaded, the average load time is "0" and not a division by zero error. This validates the logic in the relevant lambda-expression

SparkQA commented:
**[Test build #62319 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62319/consoleFull)** for PR 9571 at commit [`3971241`](https://github.com/apache/spark/commit/39712419d463bface2027fb7888f2c40f26fa412).

SparkQA commented:
**[Test build #62319 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62319/consoleFull)** for PR 9571 at commit [`3971241`](https://github.com/apache/spark/commit/39712419d463bface2027fb7888f2c40f26fa412).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62319/
Test FAILed.

steveloughran committed:
[SPARK-11373] style check in the javadocs

SparkQA commented:
**[Test build #62320 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62320/consoleFull)** for PR 9571 at commit [`0b8db8b`](https://github.com/apache/spark/commit/0b8db8b285bfcc73e861a92303206f670ea40de1).

SparkQA commented:
**[Test build #62320 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62320/consoleFull)** for PR 9571 at commit [`0b8db8b`](https://github.com/apache/spark/commit/0b8db8b285bfcc73e861a92303206f670ea40de1).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62320/
Test FAILed.

steveloughran committed:
[SPARK-11373] IDE had mysteriously re-ordered imports in the same line. Interesting, and not in a good way

steveloughran commented:
This patch adds separate average values of the load times vs merge times per event; this shows ~2x difference in replay from load in the test case.,

These `.time` gauges are little lambda expressions evaluated whenever the gauge value is extracted; they divide the total load/merge durations by the event count. The Timer metrics don't provide enough data here, because they support various decaying reservoirs/windows for their time values, not whole-life-of-app durations.

There's some base class metric support for registration (everything is prefixed) and testability. There's now tests in `HistoryServerSuite."incomplete apps get refreshed"` for metrics being in the dumped list, and for specific values, especially averages (that they increase, that they don't trigger division-by-0 exceptions before there have been any loads)

This is what the metrics look like after the tests (from a log of the toString value). 

The `.last.attempted` values are `System.currentTimeMillis` timestamps of the operations. I'd considered a "time-since" gauge, but after some offline discussion with Allen Wittenauer, went for the absolute values; I'll leave to to the management tooling to work out elapsed times
from absolute values if they want to use that for alerts or UIs.

```
16/07/14 14:17:03.837 ScalaTest-main-running-HistoryServerSuite INFO HistoryServerSuite: Metrics:
Metrics for history:
  Counters
  Gauges

Metrics for history.fs:
  Counters
    history.provider.appui.event.count = 103
    history.provider.appui.load.count = 6
    history.provider.appui.load.duration = 38244482
    history.provider.appui.load.failure.count = 0
    history.provider.appui.load.not-found.count = 0
    history.provider.history.merge.duration = 16138193
    history.provider.history.merge.event.count = 83
    history.provider.update.count = 5
    history.provider.update.failure.count = 0
  Gauges
    history.provider.appui.event.replay.time = 371305
    history.provider.history.merge.event.time = 194436
    history.provider.update.last.attempted = 1468502223726
    history.provider.update.last.succeeded = 1468502223000

Metrics for application.cache:
  Counters
    history.cache.eviction.count = 3
    history.cache.load.count = 4
    history.cache.lookup.count = 58
    history.cache.lookup.failure.count = 0
    history.cache.update.probe.count = 57
    history.cache.update.triggered.count = 3
  Gauges
```

One thing to consider: gauges of the number of complete and incomplete applications? I know the REST UI gives this, but only indirectly (you call, you count the size of the lists). Doing it in the metrics provides something that could be monitored or probed in tests without making REST calls.

SparkQA commented:
**[Test build #62321 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62321/consoleFull)** for PR 9571 at commit [`4b2046f`](https://github.com/apache/spark/commit/4b2046fd47d58bc5814f6271aa1d67a6cdd0918c).

SparkQA commented:
**[Test build #62321 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62321/consoleFull)** for PR 9571 at commit [`4b2046f`](https://github.com/apache/spark/commit/4b2046fd47d58bc5814f6271aa1d67a6cdd0918c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62321/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -667,6 +710,123 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+  /** Total time to merge all histories. */
+  val historyTotalMergeTime = new Counter()
+
+  /** Average time to load a single event in the App UI */
```
Comment looks wrong.

steveloughran replied:
replaced with " Average time to process an event in the history merge operation."

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala:
```diff
@@ -226,6 +259,135 @@ class HistoryServer(
+  override def toString: String = {
+    val sb = new StringBuilder(s"Metrics for $sourceName:\n")
+    sb.append("  Counters\n")
+    metricRegistry.getCounters.asScala.foreach { entry =>
```
Maybe use `case (name, counter) =>`?

steveloughran replied:
done

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala:
```diff
@@ -226,6 +259,135 @@ class HistoryServer(
+        sb.append("    ").append(entry._1).append(" = ").append(entry._2.getCount).append('\n')
+    }
+    sb.append("  Gauges\n")
+    metricRegistry.getGauges.asScala.foreach { entry =>
```
Same.

steveloughran replied:
done.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala:
```diff
@@ -226,6 +259,135 @@ class HistoryServer(
+   * @throws ClassCastException if the gauge is found but of the wrong type
+   */
+  def getLongGauge(gaugeName: String): Option[Gauge[Long]] = {
+    Option(metricRegistry.getGauges(new MetricFilter {
```
Can you call `getGauge` and cast instead of copy & pasting the code?

steveloughran replied:
done, &  cleaned up all the filtering to have a single `MetricByName` filter.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -667,6 +710,123 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+    ("history.merge.timer", historyMergeTimer),
+    ("appui.load.timer", appUiLoadTimer))
+
+  val allMetrics = countersAndGauges ++ timers
```
You don't seem to need these lists outside of the context of calling `register`, so can you instead inline the list into the call to avoid having these fields?

steveloughran replied:
ok. merged them all in. I'd split them by type, but like you say: not needed

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -667,6 +710,123 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+    ("appui.event.replay.time", appUIEventReplayTime)
+  )
+
+  private val timers = Seq (
```
nit: no space before `(`

steveloughran replied:
fixed

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala:
```diff
@@ -226,6 +259,135 @@ class HistoryServer(
 /**
+ * An abstract implementation of the metrics [[Source]] trait with some common operations.
+ */
+private[history] abstract class HistoryMetricSource(val prefix: String) extends Source {
```
Can you put this class (and the new gauges you're adding) in a new source file? They're not specific to `HistoryServer` so it's better to have them in a separate place.

steveloughran replied:
done; also made the`Timestamp` gauge take a Spark `Clock` for easier use/testing, plus some other standalone tests. Separate source file -> isolated tests.

These gauges are actually generic to any metrics, and the `LambdaLongGauge` gauge quite a nice one to use in a lot of places; if accompanied with `LambdaIntGauge` they could be used as a more functional equivalent to the various gauge subclasses throughout the code. I could put them somewhere more generic if you think its useful (and even, in a different patch, move the existing gauges over to them)

vanzin replied:
We can move the classes to a different module when it becomes necessary.

vanzin commented on core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala:
```diff
@@ -204,6 +204,8 @@ private[spark] class MetricsSystem private (
     }
   }
+
+  private[spark] def getMetricRegistry(): MetricRegistry = { registry }
```
nit: braces are not necessary

vanzin commented:
A few minor things left.

steveloughran committed:
[SPARK-11373] address latest comments
-nits
-recommended minor changes
+pull out lambda gauge and HistoryMetricSource into their own file, HistoryMetricSource.scala. Added tests to go with this, and made the toString call robust against failing gauges
+fixed a scaladoc warning in HistoryServer

SparkQA commented:
**[Test build #63864 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63864/consoleFull)** for PR 9571 at commit [`745d532`](https://github.com/apache/spark/commit/745d532b67b64149e86d2fdf0464cab616023eac).

SparkQA commented:
**[Test build #63864 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63864/consoleFull)** for PR 9571 at commit [`745d532`](https://github.com/apache/spark/commit/745d532b67b64149e86d2fdf0464cab616023eac).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63864/
Test PASSed.

SparkQA commented:
**[Test build #64359 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64359/consoleFull)** for PR 9571 at commit [`745d532`](https://github.com/apache/spark/commit/745d532b67b64149e86d2fdf0464cab616023eac).

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala:
```diff
@@ -225,14 +274,26 @@ class HistoryServer(
   }
 }
 
+
```
nit: too many blank lines.

steveloughran replied:
got it

vanzin commented:
Looks ok but there's one comment you haven't addressed.

SparkQA commented:
**[Test build #64359 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64359/consoleFull)** for PR 9571 at commit [`745d532`](https://github.com/apache/spark/commit/745d532b67b64149e86d2fdf0464cab616023eac).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64359/
Test FAILed.

steveloughran committed:
[SPARK-1137]  cull surplus lines

SparkQA commented:
**[Test build #64531 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64531/consoleFull)** for PR 9571 at commit [`d39996a`](https://github.com/apache/spark/commit/d39996aaabcc0de873bf410eeed6448d6d259c42).

SparkQA commented:
**[Test build #64531 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64531/consoleFull)** for PR 9571 at commit [`d39996a`](https://github.com/apache/spark/commit/d39996aaabcc0de873bf410eeed6448d6d259c42).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64531/
Test FAILed.

steveloughran commented:
Test failures timeout related; unlikely to be due to this patch
```
Test Result (2 failures / +2)
org.apache.spark.sql.hive.HiveSparkSubmitSuite.dir
org.apache.spark.sql.hive.HiveSparkSubmitSuite.dir
```

vanzin commented:
retest this please

SparkQA commented:
**[Test build #64586 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64586/consoleFull)** for PR 9571 at commit [`d39996a`](https://github.com/apache/spark/commit/d39996aaabcc0de873bf410eeed6448d6d259c42).

vanzin commented:
@steveloughran you still have not addressed my comment about `appUiLoadTimer`. The metric is still being registered even though it doesn't make any sense to the user.

SparkQA commented:
**[Test build #64586 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64586/consoleFull)** for PR 9571 at commit [`d39996a`](https://github.com/apache/spark/commit/d39996aaabcc0de873bf410eeed6448d6d259c42).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64586/
Test PASSed.

steveloughran commented:
sorry, didn't see that one. Will fix

steveloughran committed:
[SPARK-11373] don't register appui.load.timer as a metric

SparkQA commented:
**[Test build #64651 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64651/consoleFull)** for PR 9571 at commit [`1c750b5`](https://github.com/apache/spark/commit/1c750b50cca7e17ebe5950592732ff4ceae7cf2b).

SparkQA commented:
**[Test build #64651 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64651/consoleFull)** for PR 9571 at commit [`1c750b5`](https://github.com/apache/spark/commit/1c750b50cca7e17ebe5950592732ff4ceae7cf2b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64651/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -664,6 +707,116 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+
+  register(Seq(
+    ("history.merge.event.count", historyEventCount),
+    ("history.merge.event.time", historyEventMergeTime),
```
Similarly, this metric suffers from the same issue as `appUiLoadTimer` and probably shouldn't be exposed. Sorry should have pointed this out earlier.

vanzin replied:
Actually, sorry, not this one but `historyMergeTimer` below.

steveloughran replied:
No problem, though I think the total load time does need to be measured and logged somewhere. I'll make sure that happens.

steveloughran replied:
what's actually being displayed is the average time/event, so it is independent of history size, it more measures system performance during replay.

Even so, it may not be that useful, so I can cut it.

What may actually be more interesting as a remotely accessible metric is a 0/1 boolean indicating whether or not replay has completed. There currently doesn't seem to be any way to ask the history server if all histories have loaded: the UI comes up while replay is in progress, meaning that after startup/restart, a query of a history may fail not because the app doesn' t have history, but because the replay hasn't completed yet. Exposing the state to management tools allows functional tests & other scripts to block until the history is fully loaded

vanzin replied:
> I think the total load time does need to be measured

That one is fine. I'm talking about the one that measures each app separately (`history.merge.timer`); that one is not useful to the user since every app is different.

> a remotely accessible metric is a 0/1 boolean indicating whether or not replay has completed

That's good to expose, not just as a metric but also in the UI itself, if it's not currently.

steveloughran replied:
UI would be good, though it should be some separate work: it'd need to change the history provider API to add some progress check. 

FWIW the ATS history provider simply adds this as a string in the list of key-value pairs a provider can publish. Ugly and not cross-provider.

steveloughran committed:
[SPARK-11373] sync with master; tests all happy

steveloughran committed:
[SPARK-11373] sync this PR up with the master branch

SparkQA commented:
**[Test build #73695 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73695/testReport)** for PR 9571 at commit [`d8ae876`](https://github.com/apache/spark/commit/d8ae876505de9599480929905f88612dcbc3905b).

SparkQA commented:
**[Test build #73695 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73695/testReport)** for PR 9571 at commit [`d8ae876`](https://github.com/apache/spark/commit/d8ae876505de9599480929905f88612dcbc3905b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73695/
Test FAILed.

steveloughran commented:
Style police. FWIW I think the lines that failed were already >100 chars, it was just they got indented slightly more.
```
Scalastyle checks failed at following occurrences:
[error] /home/jenkins/workspace/SparkPullRequestBuilder/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:275: File line length exceeds 100 characters
[error] /home/jenkins/workspace/SparkPullRequestBuilder/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:283: File line length exceeds 100 characters
[error] /home/jenkins/workspace/SparkPullRequestBuilder/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:284: File line length exceeds 100 characters
[error] /home/jenkins/workspace/SparkPullRequestBuilder/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:498: File line length exceeds 100 characters
```
will fix

SparkQA commented:
**[Test build #75704 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75704/testReport)** for PR 9571 at commit [`ec1f2d7`](https://github.com/apache/spark/commit/ec1f2d7f8743ce6de3e83f2f9a82f1c940c8be52).

SparkQA commented:
**[Test build #75704 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75704/testReport)** for PR 9571 at commit [`ec1f2d7`](https://github.com/apache/spark/commit/ec1f2d7f8743ce6de3e83f2f9a82f1c940c8be52).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75704/
Test FAILed.

steveloughran committed:
HADOOP-11374 fix line length errors

Change-Id: I7c3fc0865abce380fcbe08b8984cfd00b3ce0faa

SparkQA commented:
**[Test build #75708 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75708/testReport)** for PR 9571 at commit [`8903dcf`](https://github.com/apache/spark/commit/8903dcfe2b927c8fc3fed9df3e9939670a016944).

SparkQA commented:
**[Test build #75708 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75708/testReport)** for PR 9571 at commit [`8903dcf`](https://github.com/apache/spark/commit/8903dcfe2b927c8fc3fed9df3e9939670a016944).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75708/
Test PASSed.

steveloughran commented:
Line lengths fixed, tests all happy.

@vanzin —any chance of adding this to your review list?

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/ApplicationCache.scala:
```diff
@@ -410,34 +409,25 @@ private[history] class CacheMetrics(prefix: String) extends Source {
 
   /** all metrics, including timers */
-  private val allMetrics = counters ++ Seq(
+  private val allMetrics: Seq[(String, Metric with Counting)] = counters ++ Seq(
```
Is declaring the type useful here in some way?

steveloughran replied:
Either I was just being explicit about what came in, or the IDE decided to get involved. Removed

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/ApplicationHistoryProvider.scala:
```diff
@@ -99,6 +104,19 @@ private[history] abstract class ApplicationHistoryProvider {
+   * Bind to the History Server: threads should be started here; exceptions may be raised
+   * Start the provider: threads should be started here; exceptions may be raised
+   * if the history provider cannot be started.
+   * The base implementation contains a re-entrancy check and should
```
This makes this interface awkward. Why can't the `HistoryServer` keep track of that?

steveloughran replied:
All the work on Yarn service model, and that of SmartFrog before it, have given me a fear of startup logic. I'll see about doing it there though

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -310,77 +338,87 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+        lastScanTime.set(newLastScanTime)
+        metrics.updateLastSucceeded.setValue(newLastScanTime)
+      } catch {
+        case e: Exception => logError(
```
`logError` goes in the next line.

steveloughran replied:
done

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -729,6 +778,116 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+    ("appui.event.count", appUIEventCount),
+    ("appui.event.replay.time", appUIEventReplayTime),
+    ("update.timer", updateTimer),
+    ("history.merge.timer", historyMergeTimer)))
```
This times is still in this list, why? It's not useful. (This is the timer I was talking about in my previous comment.)

steveloughran replied:
cut

vanzin reviewed (commented):
I was secretly hoping you'd just give up on this patch, since it will generate a lot of conflicts with the code I'm working on in parallel... but if you really want to get this in, I'd appreciate if you took the time to address the  feedback I left in previous reviews.

steveloughran commented:
> I was secretly hoping you'd just give up on this patch, since it will generate a lot of conflicts with the code I'm working on in parallel..

No. Sorry

I do suspect the reviewers of my other outstanding patches are of the same mind "if we ignore him, he'll go away". which is why I'm effectively giving up filing new patches and bug reports related to Spark.

However, that doesn't mean that I'm going to stop updating and reminding people of my current set of patches. It's easier to accept them and hope I'll go away afterwards. After all, if you'd merged this in earlier: no merge conflicts. Better all round.

vanzin commented:
> if you'd merged this in earlier: no merge conflicts

Well, if you're going to follow that route, if you had addressed feedback earlier, the patch would have been merged long ago...

steveloughran commented:
I'm going to close this PR and start one based on a reapplication of this patch onto master; gets rid of all the merge pain and is intended to be more minimal. The latest comments of this one will be applied

jiangxb1987 commented:
Could you close this PR please? @steveloughran
